# 方差分析
消费者与产品生产者、销售者或服务的提供者之间经常发生纠纷。当发生纠纷后,消费者常 常会向消协投诉。为了对几个行业的评价,消费者协会在零售业、旅游业航空公司、家电制造业分别抽取了不同的样本，其中零售业抽取7家，旅游业抽取了6家，航空公司抽取5家、家电制造业抽取了5家，然后记录了一年中消费者对总共23家服务企业投诉的次数，结果如下表。试结合Excel输出的结果分析这四个行业的服务质量是否有显著差异？

In [2]:
import pandas as pd
import numpy as np
import scipy.stats as stats

In [3]:
data = pd.read_excel("data/例10.1.xlsx",header=0)
data

零售业   旅游业  航空公司  家电制造业
0   57  68.0  31.0   44.0
1   66  39.0  49.0   51.0
2   49  29.0  21.0   65.0
3   40  45.0  34.0   77.0
4   34  56.0  40.0   58.0
5   53  51.0   NaN    NaN
6   44   NaN   NaN    NaN

## 单因素的方差分析
### 1. 提出假设
H0 : mu1 = mu2 = mu3 = mu4  
H1 : mu(1,2,3,4) 不全相等  
### 2.构造统计量

In [5]:
x_ = data.mean()
x_

零售业      49.0
旅游业      48.0
航空公司     35.0
家电制造业    59.0
dtype: float64

![x__](img/x__.jpg)

In [155]:
x__ = data.sum().sum()/len(dataList)
print(x__)
dataList = np.array(data)[pd.notnull(data)]
print(dataList)
length = []
for i in range(0,4):
    length.append(data.iloc[:,i].dropna().shape[0])
length

47.869565217391305
[57. 68. 31. 44. 66. 39. 49. 51. 49. 29. 21. 65. 40. 45. 34. 77. 34. 56.
 40. 58. 53. 51. 44.]


[7, 6, 5, 5]

In [194]:
#总平方和 ∑∑（xij - x__）^2
SST = (((dataList-x__)**2).sum())
SST

4164.608695652174

In [195]:
#组间平方和 ∑nij(x_i-x__)^2
SSA= ((x_-x__)**2*length).sum()
SSA

1456.608695652174

In [197]:
#组内平方和 ∑∑(xij- x_)^2
SSE = ((data.iloc[:,0]-49)**2).sum()+((data.iloc[:,1]-48)**2).sum()+((data.iloc[:,2]-35)**2).sum()+((data.iloc[:,3]-59)**2).sum()
SSE

2708.0

In [198]:
SST == SSA+SSE

True

In [199]:
MSA = SSA / (4 -1)
MSA

485.536231884058

In [200]:
MSE = SSE / (len(dataList)- 4)
MSE

142.52631578947367

In [204]:
F = MSA /MSE
F

3.4066426904716036

## 做出统计决策


In [206]:
F0 = stats.f.isf(0.05,3,23-4)
F0

3.127350005113399

In [212]:
F>F0

True

In [215]:
R2= SSA/SST
R2

0.34975883740838953

拒绝原假设，mu之间有显著差异

## 方差分析中的多重比较

### 1 提出假设
检验1 H0: mu1 = mu2 H1 : mu1 != mu2  
检验2 H0: mu1 = mu3 H1 : mu1 != mu3  
检验3 H0: mu1 = mu4 H1 : mu1 != mu4  
检验4 H0: mu2 = mu3 H1 : mu2 != mu3  
检验5 H0: mu2 = mu4 H1 : mu2 != mu4  
检验6 H0: mu3 = mu4 H1 : mu3 != mu4  

In [219]:
x1 = abs(x_[0] - x_[1])
print(x1)
x2 = abs(x_[0] - x_[2])
print(x2)
x3 = abs(x_[0] - x_[3])
print(x3)
x4 = abs(x_[1] - x_[2])
print(x4)
x5 = abs(x_[1] - x_[3])
print(x5)
x6 = abs(x_[2] - x_[3])
print(x6)

1.0
14.0
10.0
13.0
11.0
24.0


In [223]:
ta= stats.t.isf(0.05/2,23-4)
ta

2.0930240544082634

In [227]:
#LSD = ta/2 (MSE(1/ni+1/nj))**0.5
LSD1 = ta*(MSE*(1/length[0]+1/length[1]))**.5
print(LSD1)
LSD2 = ta*(MSE*(1/length[0]+1/length[2]))**.5
print(LSD2)
LSD3 = ta*(MSE*(1/length[0]+1/length[3]))**.5
print(LSD3)
LSD4 = ta*(MSE*(1/length[1]+1/length[2]))**.5
print(LSD4)
LSD5 = ta*(MSE*(1/length[1]+1/length[3]))**.5
print(LSD5)
LSD6 = ta*(MSE*(1/length[2]+1/length[3]))**.5
print(LSD6)

13.901727781081766
14.63114619914529
14.63114619914529
15.13064578318105
15.13064578318105
15.803444106192725


In [228]:
print(x1<LSD1)
print(x2<LSD2)
print(x3<LSD3)
print(x4<LSD4)
print(x5<LSD5)
print(x6<LSD6)

True
True
True
True
True
False


不能拒接原假设1  
不能拒接原假设2  
不能拒接原假设3  
不能拒接原假设4  
不能拒接原假设5  
拒接原假设6  航空公司和家电制造业被投诉之间有显著差异

## 双因素方差分析
### 无交互作用分析

In [99]:
df = pd.read_excel("data/例10.3.xlsx",header=1, usecols=[1,2,3,4,5,6],index_col= 0)
df

地区1  地区2  地区3  地区4  地区5
品牌1  365  350  343  340  323
品牌2  345  368  363  330  333
品牌3  358  323  353  343  308
品牌4  288  280  298  260  298

![10-1](img/10-1.jpg)

#### 构造统计量

![10-2](img/10-2.jpg)

In [135]:
k = df.shape[0]
r = df.shape[1]
data = df.values
x__ =  data.sum().sum()/(data.shape[0]*data.shape[1])
x__

328.45

In [103]:
SST = sum((data - x__)**2).sum()
SST

17888.95

![10-3](img/10-3.jpg)
![10-4](img/10-4.jpg)

In [124]:
SSR = sum(((df.sum(axis=1).values/data.shape[1] - x__)**2)*data.shape[1])
SSR

13004.549999999996

In [126]:
SSC =  sum(((df.sum()/data.shape[0] - x__)**2)*data.shape[0])
SSC

2011.7

In [127]:
SSE = SST -SSR - SSC
SSE

2872.7000000000053

![10-5](img/10-5.jpg)

In [130]:
MSR = SSR/(data.shape[0]-1)
MSR

4334.8499999999985

In [131]:
MSC = SSC/(data.shape[1]-1)
MSC

502.925

In [132]:
MSE = SSE/(data.shape[0]-1)/(data.shape[1]-1)
MSE

239.3916666666671

In [142]:
Fr = MSR/MSE
Fc = MSC/MSE
print("Fr:"+str(Fr))
print("Fc:"+str(Fc))

Fr:18.107773175061748
Fc:2.1008458941065857


In [143]:
#95%置信度下
alpha = 0.05
Far = stats.f.isf(alpha, dfn=k-1, dfd=(k-1)*(r-1))   #行 F临界值
Fac = stats.f.isf(alpha, dfn=r-1, dfd=(k-1)*(r-1))   #行 F临界值
print("Far:"+str(Far))
print("Fac:"+str(Fac))

Far:3.490294819497605
Fac:3.259166726901249


Fr>Far
所以拒绝原假设，行因素，品牌因素对销售量有明显的影响

Fc>Fac 不拒绝原假设，列因素，不能认为地区对销售量有影响

In [147]:
pfr = stats.f.sf(Fr, k-1, (k-1)*(r-1))  # 行变量 统计量F的P值
pfc = stats.f.sf(Fc, r-1, (k-1)*(r-1))  # 列变量 统计量F的P值
print("pfr:"+str(pfr))
print("pfc:"+str(pfc))

pfr:9.456152887730434e-05
pfc:0.14366488731130558


pfr<0.05  
pfr>0.05 

In [149]:
RR = (SSR+SSC)/SST
RR

0.8394148342971497

In [151]:
R = RR**.5
R

0.9161958493123343

品牌和地区合起来总共解释了销售差异的83.94%，R表明两者和销售量有明显关系
#### 总结
![10-6](img/10-6.jpg)
### 有交互作用的方差分析
城市道路交通管理部门为了研究不同的路段和不同的时间段对行车时间的影响，,让一名交通警察分别在两个路段和高峰期 与非高峰期亲自驾车进行试验,通过试验共获得 20个行车时间(单位:m:分钟)的数据。试分析影响a=0.05
![10-7](img/10-7.jpg)
![10-8](img/10-8.jpg)
![10-9](img/10-9.jpg)

In [166]:
df = pd.read_excel("data/例10.5.xlsx",header=0,index_col=0)
df.head()

路段1  路段2
1   26   19
2   24   20
3   27   23
4   25   22
5   25   21

In [287]:
k = 2
r = 2
m = 5
n = df.shape[0]*df.shape[1]
x__ =  df.values.sum()/n
x__

20.25

In [288]:
SST = sum(( df.values - x__)**2).sum()
SST

329.75

In [279]:
d1 = df[0:5]
d2 = df[5:]
x_i = [d1.mean().mean(),d2.mean().mean()]
print(x_i)
SSR = sum((x_i-x__)**2).sum()*r*m
SSR

[23.2, 17.299999999999997]


174.05000000000013

In [280]:
x_j = df.mean(axis=0).values
x_j

array([22.4, 18.1])

In [281]:
SSC = sum((x_j-x__)**2).sum()*k*m
SSC

92.44999999999989

In [282]:
x_ = df.values.mean()
x_

20.25

In [283]:
x_ij = np.array([d1.mean().values,d2.mean().values])
x = x_ij+x__

In [284]:
SSRC = ((x[0,0]-x_i[0]-x_j[0])**2 + (x[0,1]-x_i[0]-x_j[1])**2+ (x[1,1]-x_i[1]-x_j[1])**2+ (x[1,0]-x_i[1]-x_j[0])**2)*m
SSRC

0.050000000000001425

In [297]:
SSE  = SST-SSR-SSC -SSRC
SSE

63.19999999999998

In [345]:
MSR = SSR/(k-1)
MSE = SSE/(m-1)/k/r
MSC = SSC/1
Fr = MSR/MSE
Fc = MSC/MSE
FRC = SSRC/1/MSE
print("Fr:"+str(Fr))
print("Fc:"+str(Fc))
print("FRC:"+str(FRC))

Fr:44.06329113924055
Fc:23.405063291139218
FRC:0.01265822784810163


In [346]:
alpha = 0.05
pfr = stats.f.sf(Fr, 1,m-1)   #行 F临界值
print("pfr:"+str(pfr))
pfc = stats.f.sf(Fc, 1,m-1)   #行 F临界值
print("pfc:"+str(pfc))
pfrc = stats.f.sf(FRC, 1,m-1)   #行 F临界值
print("pfrc:"+str(pfrc))

pfr:0.0026729375641445672
pfc:0.00841302296398786
pfrc:0.9158401958222961


In [347]:
print(pfr<0.05)
print(pfc<0.05)
print(pfrc<0.05)

True
True
False


拒绝原假设，不同时间段有显著差异  
拒绝原假设，不同路段对行车时间有显著差异  
不拒绝原假设，没有证据表明时段和路段对行车时间有显著差异